src/neo4j_genai/retrievers/base.py

```diff
@@ -13,15 +13,70 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 from __future__ import annotations
-from abc import ABC, abstractmethod
-from typing import Optional, Callable, Any
+import types
+import inspect
+from abc import ABC, abstractmethod, ABCMeta
+from typing import Optional, Callable, Any, TypeVar
+
+try:
+    from typing import ParamSpec
+except ImportError:
+    from typing_extensions import ParamSpec  # type: ignore
+
 import neo4j
 
 from neo4j_genai.types import RawSearchResult, RetrieverResult, RetrieverResultItem
 from neo4j_genai.exceptions import Neo4jVersionError
 
+T = ParamSpec("T")
+P = TypeVar("P")
+
 
-class Retriever(ABC):
+def copy_function(f: Callable[T, P]) -> Callable[T, P]:
```

Comment:
This is a cool approach but I'm wondering whether doing a deepcopy for every retriever instance just to get typehints and docstrings is a bit too complex, and whether it introduces any performance overhead from the deep copying. 

Did you consider other approaches that don't involve copying? Something like

```python
def propagate_docstring_signature(cls):
    get_search_results = getattr(cls, "get_search_results", None)
    if get_search_results and "search" not in cls.__dict__:

        def search(self, *args, **kwargs):
            return get_search_results(self, *args, **kwargs)

        search.__doc__ = get_search_results.__doc__
        search.__signature__ = inspect.signature(get_search_results)
        setattr(cls, "search", search)
    return cls


@propagate_docstring_signature
class Retriever(ABC):
    ...
```

Inspired by this post: https://stackoverflow.com/questions/147816/preserving-signatures-of-decorated-functions

Reply:
Unless there was a specific reason to go with this approach?

Reply:
I agree about the bad complexity/gain ratio, but couldn't find any other working solution that do not require more work for the developer. 
If not (deep) copying, the search method is always the one attached to the BaseRetriever, and hence gets the last documentation attached to it, for all retrievers. In your code snippet, we have to consider that the search method is more complex than just calling the `get_search_results` method, and it's also a bit weird to me to have this method "hanging" in the middle.

Regarding performances, this is done only once when the class is discovered (so on first import of the package), so it's not a concern IMHO.

The only other working solution I could think of, is to override the search method with a docstring and parameters. The advantage is that we would have autocompletion working, but then the developer needs to instantiate this method "just" for documentation purposes.

Reply:
I see, in that case I'm happy to go with this approach 👍